File: inflearn-algorithm-class/2023-2 알고리즘/정렬 알고리즘/bubbleSort.py
import sys 

def bubbleSorts(arr):
    ans1 = bubbleSort1(arr[:])
    ans2 = bubbleSort2(arr[:])
    ans3 = bubbleSort3(arr[:])
    print(ans1[0], ans1[1], ans2[0], ans2[1], ans3[0], ans3[1])

def bubbleSort1(arr):
    calculation_count, swap_count = 0, 0
    size = arr.pop(0)

    for i in range(1,size):
        for j in range(1, size-i+1):
            calculation_count += 1
            if arr[j-1] > arr[j]:
                (arr[j-1], arr[j]) = (arr[j], arr[j-1])
                swap_count += 1

    return [calculation_count, swap_count]

def bubbleSort2(arr):
    calculation_count, swap_count = 0, 0
    size = arr.pop(0)

    for i in range(1,size):
        swapped = False
        for j in range(1, size-i+1):
            calculation_count += 1
            if arr[j-1] > arr[j]:
                (arr[j-1], arr[j]) = (arr[j], arr[j-1])
                swapped = True
                swap_count += 1

        if swapped == False:
            break

    return [calculation_count, swap_count]

def bubbleSort3(arr):
    calculation_count, swap_count = 0, 0
    size = arr.pop(0)
    lastSwappedIndex = size

    while lastSwappedIndex > 0:
        swappedIndex = 0
        for j in range(1, lastSwappedIndex):
            calculation_count += 1
            if arr[j-1] > arr[j]:
                (arr[j-1], arr[j]) = (arr[j], arr[j-1])
                swappedIndex = j
                swap_count += 1
        lastSwappedIndex = swappedIndex

    return [calculation_count, swap_count]


n = int(sys.stdin.readline());

for i in range(0, n):
    num_list = list(map(int, sys.stdin.readline().split()));
    bubbleSorts(num_list)

File: inflearn-algorithm-class/2023-2 알고리즘/정렬 알고리즘/bubbleSort.cpp
#include <iostream>
#include <vector>
using namespace std;

vector<int> bubbleSort1(vector<int> arr) {
    vector<int> result(2, 0); 
    int size = arr[0];
    arr.erase(arr.begin());

    for (int i = 1; i < size; i++) {
        for (int j = 1; j < size - i + 1; j++) {
            result[0]++; 
            if (arr[j - 1] > arr[j]) {
                swap(arr[j - 1], arr[j]);
                result[1]++; 
            }
        }
    }

    return result;
}

vector<int> bubbleSort2(vector<int> arr) {
    vector<int> result(2, 0); 
    int size = arr[0];
    arr.erase(arr.begin());

    for (int i = 1; i < size; i++) {
        bool swapped = false;
        for (int j = 1; j < size - i + 1; j++) {
            result[0]++; 
            if (arr[j - 1] > arr[j]) {
                swap(arr[j - 1], arr[j]);
                swapped = true;
                result[1]++; 
            }
        }
        if (!swapped) {
            break;
        }
    }

    return result;
}

vector<int> bubbleSort3(vector<int> arr) {
    vector<int> result(2, 0); 
    int size = arr[0];
    arr.erase(arr.begin());
    int lastSwappedIndex = size;

    while (lastSwappedIndex > 0) {
        int swappedIndex = 0;
        for (int j = 1; j < lastSwappedIndex; j++) {
            result[0]++; 
            if (arr[j - 1] > arr[j]) {
                swap(arr[j - 1], arr[j]);
                swappedIndex = j;
                result[1]++; 
            }
        }
        lastSwappedIndex = swappedIndex;
    }

    return result;
}

int main() {
    int n;
    cin >> n;

    for (int i = 0; i < n; i++) {
        int size;
        cin >> size;
        vector<int> num_list(size + 1);
        num_list[0] = size;

        for (int j = 1; j <= size; j++) {
            cin >> num_list[j];
        }

        vector<int> ans1 = bubbleSort1(num_list);
        vector<int> ans2 = bubbleSort2(num_list);
        vector<int> ans3 = bubbleSort3(num_list);

        cout << ans1[0] << " " << ans1[1] << " " << ans2[0] << " " << ans2[1] << " " << ans3[0] << " " << ans3[1] << endl;
    }

    return 0;
}
